File: Server~/src/tools/createMaterialTool.ts
import * as z from 'zod';
import { Logger } from '../utils/logger.js';
import { McpUnity } from '../unity/mcpUnity.js';
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { McpUnityError, ErrorType } from '../utils/errors.js';
import { CallToolResult } from '@modelcontextprotocol/sdk/types.js';

// Constants for the tool
const toolName = 'create_material';
const toolDescription = "Creates a new material asset at the given path, using the named shader (defaults to " +
  "'Standard' if not found)";
const paramsSchema = z.object({
  path: z.string().describe('Assets path to save the new material to (e.g. "Assets/Materials/Red.mat")'),
  shaderName: z.string().optional().describe('Name of the shader to use (e.g. "Standard", "Universal Render Pipeline/Lit")')
});

/**
 * Creates and registers the Create Material tool with the MCP server
 * This tool allows creating a new material asset in the Unity project
 *
 * @param server The MCP server instance to register with
 * @param mcpUnity The McpUnity instance to communicate with Unity
 * @param logger The logger instance for diagnostic information
 */
export function registerCreateMaterialTool(server: McpServer, mcpUnity: McpUnity, logger: Logger) {
  logger.info(`Registering tool: ${toolName}`);

  server.tool(
    toolName,
    toolDescription,
    paramsSchema.shape,
    async (params: any) => {
      try {
        logger.info(`Executing tool: ${toolName}`, params);
        const result = await toolHandler(mcpUnity, params);
        logger.info(`Tool execution successful: ${toolName}`);
        return result;
      } catch (error) {
        logger.error(`Tool execution failed: ${toolName}`, error);
        throw error;
      }
    }
  );
}

/**
 * Handles creating a material asset in Unity
 *
 * @param mcpUnity The McpUnity instance to communicate with Unity
 * @param params The parameters for the tool
 * @returns A promise that resolves to the tool execution result
 * @throws McpUnityError if the request to Unity fails
 */
async function toolHandler(mcpUnity: McpUnity, params: any): Promise<CallToolResult> {
  if (!params.path || params.path.trim() === '') {
    throw new McpUnityError(
      ErrorType.VALIDATION,
      "Required parameter 'path' not provided"
    );
  }

  const response = await mcpUnity.sendRequest({
    method: toolName,
    params
  });

  if (!response.success) {
    throw new McpUnityError(
      ErrorType.TOOL_EXECUTION,
      response.message || `Failed to create material`
    );
  }

  return {
    content: [{
      type: response.type,
      text: response.message || `Successfully created material`
    }]
  };
}

File: Editor/Tools/CreateMaterialTool.cs
using UnityEngine;
using UnityEditor;
using McpUnity.Unity;
using McpUnity.Utils;
using Newtonsoft.Json.Linq;

namespace McpUnity.Tools
{
    /// <summary>
    /// Tool for creating a new material asset with a given shader
    /// </summary>
    public class CreateMaterialTool : McpToolBase
    {
        private const string DefaultShaderName = "Standard";

        public CreateMaterialTool()
        {
            Name = "create_material";
            Description = "Creates a new material asset at the given path, using the named shader (defaults to 'Standard' if not found)";
        }

        /// <summary>
        /// Execute the CreateMaterial tool with the provided parameters synchronously
        /// </summary>
        /// <param name="parameters">Tool parameters as a JObject. Expects 'path' and optional 'shaderName'</param>
        public override JObject Execute(JObject parameters)
        {
            string path = parameters["path"]?.ToObject<string>();
            string shaderName = parameters["shaderName"]?.ToObject<string>();

            if (string.IsNullOrEmpty(path))
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    "Required parameter 'path' not provided",
                    "validation_error"
                );
            }

            if (!path.StartsWith("Assets/") || !path.EndsWith(".mat"))
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    $"Invalid material path '{path}'. Path must start with 'Assets/' and end with '.mat'.",
                    "validation_error"
                );
            }

            if (AssetDatabase.LoadAssetAtPath<Material>(path) != null)
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    $"A material already exists at '{path}'.",
                    "validation_error"
                );
            }

            Shader shader = string.IsNullOrEmpty(shaderName) ? null : Shader.Find(shaderName);
            string resolvedShaderName = shaderName;
            if (shader == null)
            {
                shader = Shader.Find(DefaultShaderName);
                resolvedShaderName = DefaultShaderName;
            }

            if (shader == null)
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    $"Could not find shader '{shaderName ?? DefaultShaderName}' to create the material.",
                    "not_found_error"
                );
            }

            Material material = new Material(shader);
            AssetDatabase.CreateAsset(material, path);

            McpLogger.LogInfo($"[MCP Unity] Created material '{path}' with shader '{resolvedShaderName}'");

            return new JObject
            {
                ["success"] = true,
                ["type"] = "text",
                ["message"] = $"Successfully created material '{path}' with shader '{resolvedShaderName}'",
                ["path"] = path,
                ["shaderName"] = resolvedShaderName
            };
        }
    }
}
